account_mtd/models/mtd_hello_world.py

```diff
@@ -0,0 +1,406 @@
+            "{}&redirect_uri={}".format(state, redirect_uri)
+        )
+        self._logger.info(
+            "(Step 1) Get authorisation - authorisation URI " +
+            "used to send request:- {}".format(authorisation_url)
+        )
+        req = requests.get(authorisation_url, timeout=3)
+        #response_token = json.loads(req.text)
+        self._logger.info(
+            "(Step 1) Get authorisation - received response of the request:- {}".format(req)
+        )
+        if req.ok:
+            return {'url': authorisation_url, 'type': 'ir.actions.act_url', 'target': 'self', 'res_id': self.id}
+        else:
+            response_token = json.loads(req.text)
+            construct_text = (
+                    "Date {}     Time {} \n".format(datetime.now().date(), datetime.now().time()) +
+                    "Sorry. The connection failed ! \n" +
+                    "Please check the log below for details. \n\n" +
+                    "Connection Status Details: \n" +
+                    "Request Sent: \n{} \n\n".format(authorisation_url) +
+                    "Error Code:\n{} \n\n".format(req.status_code) +
+                    "Response Received: \n{}".format(response_token['message'])
+            )
+            self.response_from_hmrc = construct_text
+            # We need to set the response received in tracker to be True 
+            # even if we receive negetive response.
+            tracker_api.response_received = True
+            action = self.env.ref('account_mtd.action_mtd_hello_world')
+            menu_id = self.env.ref('account_mtd.submenu_mtd_hello_world')
+            return werkzeug.utils.redirect(
+                '/web#id={}&view_type=form&model=mtd.hello_world'.format(self.id) +
+                '&menu_id={}&action={}'.format(menu_id.id, action.id))
+            
+    @api.multi        
+    def exchange_user_authorisation(self, auth_code, record_id, tracker_id):
+        self._logger.info("(Step 2) exchange authorisation code")
+        api_tracker = self.env['mtd.api_request_tracker'].search([('id', '=', tracker_id)])
+        api_token = self.env['mtd.api_tokens'].search([('api_id', '=', api_tracker.api_id)])
+
+        if api_token:
+            api_token.authorisation_code = auth_code
+        else:
+            api_token.create({
+                'api_id': api_tracker.api_id,
+                'api_name': api_tracker.api_name,
+                'authorisation_code': auth_code
+                })
+        # self.current_record = record_id
+        record = self.env['mtd.hello_world'].search([('id', '=', record_id)])
+        token_location_uri = "https://test-api.service.hmrc.gov.uk/oauth/token"
+        client_id = record.hmrc_configuration.client_id
+        client_secret = record.hmrc_configuration.client_secret
+        redirect_uri = "http://localhost:8090{}".format('/auth-redirect')
+
+        data_user_info = {
```

Comment:
What is this for?

Reply:
the data user info is used to to pass the parameters in a post request